### ISSUE TITLE
Repeatedly importing files with many notations for the same head constant is slow

### ISSUE THREAD

**Issue by Janno** [coq/coq#19357]
### Description of the problem

This was discovered by @ppedrot during CUDW24. @ppedrot, do you still have the reproducing example? If not, I can make a new one. 

IIRC the problem is that the code uses a list ordered by recency; one per head constant. When a file with many such notations is re-imported, the notations need to be inserted in the front and removed from the tail of the list. I think the removal happens one by one and needs to traverse the entire list in the worst case. I guess this an instance of accidentally quadratic complexity? (I can't find the code right now so I might have be misremembering the detais.)

### Version of Coq where this bug occurs

8.19, master

**Comment by SkySkimmer** [comment]
Is the example something like
~~~coq
Module Source.
(* keep compact representations *)
Open Scope dec_uint_scope.

Axiom foo : Number.uint -> Prop.

Notation "'#0'" := (foo 0).
Notation "'#1'" := (foo 1).
Notation "'#2'" := (foo 2).
Notation "'#3'" := (foo 3).
Notation "'#4'" := (foo 4).
Notation "'#5'" := (foo 5).
Notation "'#6'" := (foo 6).
Notation "'#7'" := (foo 7).
Notation "'#8'" := (foo 8).
Notation "'#9'" := (foo 9).
Notation "'#10'" := (foo 10).
(* etc *)

End Source.

Time Import Source.
Time Import Source.
Time Import Source. (* same results as the 2nd import *)
~~~
?

With 2k notations the first import takes 0.25s and importing again takes 0.45s.
If we use nat instead of uint (decimal) compiling the notations takes a while, then the first import takes 14s and importing again takes 29s.

**Comment by Janno** [comment]
Yeah the example looked like that. I am not sure why unary versus decimal makes a difference. That could be a different issue.

**Comment by SkySkimmer** [comment]
Unary being bad is not surprising IMO

**Comment by ppedrot** [comment]
FTR the problem lies in Notationextern.keymap_add, as we're repeatedly inserting notations in a list in O(n):
```ocaml
let keymap_add key interp map =
  let old = try KeyMap.find key map with Not_found -> [] in
  (* strictly finer interpretation are kept in front *)
  let strictly_finer, rest = List.partition (fun c -> strictly_finer_interpretation_than c interp) old in
  KeyMap.add key (strictly_finer @ interp :: rest) map
```

**Comment by Janno** [comment]
Is `old` totally ordered by some order compatible with `strictly_finer_interpretation_than`? I tried reading the code but I have no intuition for how exactly the sorting works right now. If it was sorted that way, we could at least avoid traversing the tail end of the list once we find the correct spot for `interp`.

Under the same assumption one could make the insertion lazy in the hopes that it can be batched. But I don't know enough about this part of the code to say if there are consecutive calls without `keymap_find` between them.

**Comment by ppedrot** [comment]
> Under the same assumption one could make the insertion lazy

I've written a patch to make the insertion lazy, you may want to try it if you want:

https://github.com/ppedrot/coq/tree/delay-notation-import

IIRC in your case you have a default Oth key for all of your notations so you're going to force the set very quickly, and it's thus likely not to make a lot of difference.

> Is old totally ordered by some order compatible with strictly_finer_interpretation_than

The code is completely nonsensical, the order is not total so this is just some broken set approximation. I think we could try to do something more reasonable by having a dnet of notation patterns instead, but I have a hard time wrapping my head around recursive notations to really understand the desired spec of this piece of code.

**Comment by Janno** [comment]
This is definitely a noticeable improvement:


| Relative | Master   | MR       | Change   | Filename
|---------:|---------:|---------:|---------:|----------
|   -1.35% |  35977.3 |  35490.2 |   -487.1 | cpp2v-generated
|   -1.06% |  88438.4 |  87498.2 |   -940.2 | other
|   -1.15% | 124415.7 | 122988.4 |  -1427.3 | total

Apart from the main files that contain one huge `Definition`, the "cpp2v-generated" category also includes the notation files (which we looked at during CUDW) that declare a whole lot of somewhat overlapping notation. Some of those files are 80% faster(!). The impact on "other" (proofs) seems to be mostly from `Import` and `Export`s. Some files re-export notations defined in specific "cpp2v-generated" files to make them available in the final proof files. Those `Export` calls went down by >90%, in some cases from ~0.5s to less than 0.05s.

Overall this change would save about 4 minutes of computation time on our code base for normal builds. (Full builds, which include generated files for libraries that are not being verified, will probably benefit even more but I haven't tested that yet.)

**Comment by Janno** [comment]
Thanks for opening the pull request. You mention in the PR that the order is in fact not total. And the code uses `List.partition` instead of binary search. I think this means one could craft a new notation entry that switches the relative order of existing entries. Is that correct? If so, that might actually warrant a new issue. (I don't understand why the order is not total so I couldn't come up with an example.)

**Comment by ppedrot** [comment]
> Is that correct?

Yes.

> I don't understand why the order is not total.

The order is trying to capture the fact that a pattern is included in another one, seen as trees. If you consider e.g. a binary symbol `foo` then for any terms `t` and `u`, `foo t u` ≤ `foo _ u` and `foo t u` ≤ `foo t _` but `foo t _` and `foo _ u` are incomparable. Hence the proposal to use a dnet instead.
